### PR TITLE
slightly reorder boolean members to reduce struct sizes

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Slightly reorder instance members so boolean values are placed adjacently and make structs smaller.
Also make several functions const or noexcept.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is a trivial rework / code cleanup without any test coverage.
However, if something is terribly wrong, the AQL tests should reveal it.

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5786/

### CODE REVIEW

File: arangod/Aql/DocumentProducingHelper.h [hunk 1]
```diff
@@ -177,23 +178,22 @@ struct DocumentProducingFunctionContext {
...
   /// @brief Flag if the current index pointer is the last of the list.
   ///        Used in uniqueness checks.
   bool _isLastIndex;
 
   /// @brief Flag if we need to check for uniqueness
   bool _checkUniqueness;
+  size_t _numScanned;
+
+  /// @brief set of already returned documents. Used to make the result distinct
+  std::unordered_set<TRI_voc_rid_t> _alreadyReturned;
```
Comment (dothebart) [inline arangod/Aql/DocumentProducingHelper.h]: shouldn't this be 
std::unord... 
size_t ...
bool ... 
here?

Comment (jsteemann) [reply arangod/Aql/DocumentProducingHelper.h]: why?
the crucial thing is that the boolean values are placed adjacently, as they can be 1-byte aligned.
all other members have an alignment of 8 bytes.

Comment (dothebart) [reply arangod/Aql/DocumentProducingHelper.h]: ok, in this case its 5x bool, so the slack-space will be used anyways, however, I'd prefer to have them at the end, so its a non-brainer to add more there.

Comment (jsteemann) [reply arangod/Aql/DocumentProducingHelper.h]: You can have 5x anywhere, and you can add more.
The effect will be exactly the same.

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5790/

File: arangod/Aql/EnumerateCollectionExecutor.h [hunk 2]
```diff
@@ -63,26 +63,25 @@ class EnumerateCollectionExecutorInfos : public ExecutorInfos {
...
-  bool getRandom() { return _random; };
-  RegisterId getOutputRegisterId() { return _outputRegisterId; };
+  }
+  bool getProduceResult() const { return _produceResult; }
+  bool getUseRawDocumentPointers() const { return _useRawDocumentPointers; }
+  bool getRandom() const { return _random; }
+  RegisterId getOutputRegisterId() const { return _outputRegisterId; }
 
  private:
   RegisterId _outputRegisterId;
```
Comment (mpoeter) [inline arangod/Aql/EnumerateCollectionExecutor.h]: `RegisterId` is an `unsigned int` (i.e., usually 32bit), so it should be moved before the bools.

File: arangod/Aql/DocumentProducingHelper.h [hunk 3]
```diff
@@ -177,23 +178,22 @@ struct DocumentProducingFunctionContext {
   InputAqlItemRow const& _inputRow;
   OutputAqlItemRow* _outputRow;
   RegisterId const _outputRegister;
```
Comment (mpoeter) [inline arangod/Aql/DocumentProducingHelper.h]: RegisterId is an unsigned int (i.e., usually 32bit), so it should be moved before the bools.

File: arangod/Aql/IndexExecutor.h [hunk 4]
```diff
@@ -138,13 +134,17 @@ class IndexExecutorInfos : public ExecutorInfos {
...
   std::vector<std::string> const& _projections;
+  std::vector<size_t> const& _coveringIndexAttributePositions;
   transaction::Methods* _trxPtr;
   std::vector<Variable const*> _expInVars;  // input variables for expresseion
   std::vector<RegisterId> _expInRegs;       // input registers for expression
 
-  std::vector<size_t> const& _coveringIndexAttributePositions;
-  bool _useRawDocumentPointers;
   std::vector<std::unique_ptr<NonConstExpression>> _nonConstExpression;
+  /// @brief true if one of the indexes uses more than one expanded attribute,
```
Comment (mpoeter) [inline arangod/Aql/IndexExecutor.h]: Line 132 contains another `RegisterId` that should be moved here.

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5801/

**Pull request merged**